commercetools/resource_tax_category.go

```diff
@@ -0,0 +1,393 @@
+		for i, rate := range ctType.Rates {
+			rateData := make(map[string]interface{})
+
+			subRateData := make([]map[string]interface{}, len(rate.SubRates))
+			for srIndex, subrate := range rate.SubRates {
+				subRateData[srIndex] = map[string]interface{}{
+					"name":   subrate.Name,
+					"amount": subrate.Amount,
+				}
+			}
+
+			rateData["id"] = rate.ID
+			rateData["name"] = rate.Name
+			rateData["amount"] = rate.Amount
+			rateData["included_in_price"] = rate.IncludedInPrice
+			rateData["country"] = rate.Country
+			rateData["state"] = rate.State
+			rateData["sub_rate"] = subRateData
+
+			taxRates[i] = rateData
+		}
+
+		d.Set("version", ctType.Version)
+		d.Set("key", ctType.Key)
+		d.Set("name", ctType.Name)
+		d.Set("description", ctType.Description)
+		d.Set("rate", taxRates)
+	}
+	return nil
+}
+
+func resourceTaxCategoryUpdate(d *schema.ResourceData, m interface{}) error {
+	svc := getTaxCategoryService(m)
+
+	input := &taxcategories.UpdateInput{
+		ID:      d.Id(),
+		Version: d.Get("version").(int),
+		Actions: commercetools.UpdateActions{},
+	}
+
+	if d.HasChange("name") {
+		newName := d.Get("name").(string)
+		input.Actions = append(
+			input.Actions,
+			&taxcategories.ChangeName{Name: newName})
+	}
+
+	if d.HasChange("key") {
+		newKey := d.Get("key").(string)
+		input.Actions = append(
+			input.Actions,
+			&taxcategories.SetKey{Key: newKey})
+	}
+
+	if d.HasChange("description") {
+		newDescription := d.Get("description").(string)
+		input.Actions = append(
+			input.Actions,
+			&taxcategories.SetDescription{Description: newDescription})
+	}
+
+	if d.HasChange("rate") {
+		old, new := d.GetChange("rate")
+		rateChangeActions, err := resourceTaxCategoryRateChangeActions(
+			old.([]interface{}), new.([]interface{}))
+		if err != nil {
+			return err
+		}
+		input.Actions = append(input.Actions, rateChangeActions...)
+	}
+
+	log.Printf(
+		"[DEBUG] Will perform update operation with the following actions:\n%s",
+		stringFormatActions(input.Actions))
+
+	_, err := svc.Update(input)
+	if err != nil {
+		if ctErr, ok := err.(commercetools.Error); ok {
+			log.Printf("[DEBUG] %v: %v", ctErr, stringFormatErrorExtras(ctErr))
+		}
+		return err
+	}
+
+	return resourceTaxCategoryRead(d, m)
+}
+
+func resourceTaxCategoryRateChangeActions(oldValues []interface{}, newValues []interface{}) ([]commercetools.UpdateAction, error) {
+	oldLookup := createLookup(oldValues, "name")
+	newLookup := createLookup(newValues, "name")
+	newAttrDefinitions := []taxcategories.TaxRate{}
+	actions := []commercetools.UpdateAction{}
+
+	for name, value := range oldLookup {
+		if _, ok := newLookup[name]; !ok {
+			oldV := value.(map[string]interface{})
+			log.Printf("[DEBUG] Tax Rate deleted: %s", name)
+			id := oldV["id"].(string)
+			actions = append(actions, taxcategories.RemoveTaxRate{TaxRateID: id})
+		}
+	}
+
+	for name, value := range newLookup {
+		oldValue, existingField := oldLookup[name]
+		newV := value.(map[string]interface{})
+
```

Comment:
The structure of this code was a little bit hard to follow for me; perhaps the taxRateDraft construction should only happen if it's not an existingField? And the taxRate construction should only happen if the reflect.DeepEqual is not true?